### PR TITLE
 impl burning of gas, add caller_token to AccountMeta of Cancel operation

### PR DESCRIPTION
**Pull request by sinev-valentine**
When cancel operation is processed it is need to transfer all used gas to operator

Dependency: https://github.com/neonlabsorg/proxy-model.py/pull/178

### CODE REVIEW

File: evm_loader/program/src/entrypoint.rs [hunk 1]
```diff
@@ -527,21 +527,25 @@ fn process_instruction<'a>(
             let storage_info = next_account_info(account_info_iter)?;
 
             let operator_sol_info = next_account_info(account_info_iter)?;
+            let operator_eth_info = next_account_info(account_info_iter)?;
             let incinerator_info = next_account_info(account_info_iter)?;
```
Comment (otselnik) [inline evm_loader/program/src/entrypoint.rs:531]: Caller token account should be mentioned here like user_eth_info or other instructions should be rebuilt.

File: evm_loader/program/src/entrypoint.rs [hunk 2]
```diff
@@ -556,6 +560,33 @@ fn process_instruction<'a>(
...
+
+            let (gas_limit, gas_price) = storage.get_gas_params()?;
+            if used_gas > gas_limit {
+                return Err!(ProgramError::InvalidArgument);
+            }
+            let gas_price_wei = U256::from(gas_price);
+            let fee = U256::from(used_gas)
+                .checked_mul(gas_price_wei).ok_or_else(||E!(ProgramError::InvalidArgument))?;
+
+            let (caller_info, caller_token_info) = account_storage.get_caller_account_info().ok_or_else(||E!(ProgramError::InvalidArgument))?;
```
Comment (otselnik) [inline evm_loader/program/src/entrypoint.rs:580]: Why user token is being getting here, but not mentioned in accounts like in `Continue` or `CallFromRawEthereumTX`

File: evm_loader/program/src/account_storage.rs [hunk 3]
```diff
@@ -221,10 +221,10 @@ impl<'a> ProgramAccountStorage<'a> {
     }
 
     /// Get caller account info
-    pub fn get_caller_account_info(&self) -> Option<&AccountInfo<'a>> {
+    pub fn get_caller_account_info(&self) -> Option<(&AccountInfo<'a>, &AccountInfo<'a>)> {
         if let Some(account_index) = self.find_account(&self.origin()) {
-            let AccountMeta{ account, token: _, code: _ } = &self.account_metas[account_index];
-            return Some(account);
+            let AccountMeta{ account, token, code: _ } = &self.account_metas[account_index];
+            return Some((account, token));
```
Comment (otselnik) [inline evm_loader/program/src/account_storage.rs:227]: Do we really need token account to be returned?

**Comment by otselnik** [comment]
Caller token for payment for transaction should be passed in first part of accounts list, and there is no need for returning token account in `get_caller_account_info`.
Or all instructions, like `Continue` and `CallFromRawEthereumTX` should be rebuilt to use token account from `get_caller_account_info`.

File: evm_loader/program/src/entrypoint.rs (hunk 1 above)
Comment (sinev-valentine) [reply evm_loader/program/src/entrypoint.rs:532]: fixed

File: evm_loader/program/src/entrypoint.rs (hunk 2 above)
Comment (sinev-valentine) [reply evm_loader/program/src/entrypoint.rs]: fixed

File: evm_loader/program/src/account_storage.rs (hunk 3 above)
Comment (sinev-valentine) [reply evm_loader/program/src/account_storage.rs]: token was removed

**Review by s-medvedev — DISMISSED** [review]

**Review by s-medvedev — APPROVED** [review]

**Pull request merged**